File: src/py-dev/LiENa/LiENaRealTimeTask/LienaHeartbeatTask.py
# -*- coding: utf-8 -*-
import threading
import time
from LiENaBasic.lienaDefinition import *
from LiENaStructure.LiENaDatagram.LienaDatagram import LienaDatagram
from LiENaStructure.LiENaMessage.LienaHeartbeatMessage import LienaHeartbeatMessage


class LienaHearBeatTask:
    def __init__(self, output_queue, global_parameter):
        self.output_queue = output_queue
        self.global_parameter = global_parameter

        self.outputMessageCache = None
        self.rtPeriod = 1
        self.stand_by = False

        self.flag = None
        self.heartbeat_thread = threading.Thread(None, self.heartbeat)

    def heartbeat(self):
        while self.flag:

            if self.stand_by:
                time.sleep(1)
                continue

            self.send_heartbeat_message()
            time.sleep(self.rtPeriod)

    def encode_heartbeat_message(self, message):
        bytes_to_send = bytearray(self.global_parameter.get_global_datagram_size())

        data_type_msb = message.get_message_id() // (2 ** 32)
        data_type_lsb = message.get_message_id() % (2 ** 32)

        bytes_to_send[0] = (data_type_msb & 0xff000000) >> 24
        bytes_to_send[1] = (data_type_msb & 0x00ff0000) >> 16
        bytes_to_send[2] = (data_type_msb & 0x0000ff00) >> 8
        bytes_to_send[3] = (data_type_msb & 0x000000ff)
        bytes_to_send[4] = (data_type_lsb & 0xff000000) >> 24
        bytes_to_send[5] = (data_type_lsb & 0x00ff0000) >> 16
        bytes_to_send[6] = (data_type_lsb & 0x0000ff00) >> 8
        bytes_to_send[7] = (data_type_lsb & 0x000000ff)

        bytes_to_send[8] = (message.get_target_id() & 0xff000000) >> 24
        bytes_to_send[9] = (message.get_target_id() & 0x00ff0000) >> 16
        bytes_to_send[10] = (message.get_target_id() & 0x0000ff00) >> 8
        bytes_to_send[11] = (message.get_target_id() & 0x000000ff)

        bytes_to_send[12] = (message.get_origin_id() & 0xff000000) >> 24
        bytes_to_send[13] = (message.get_origin_id() & 0x00ff0000) >> 16
        bytes_to_send[14] = (message.get_origin_id() & 0x0000ff00) >> 8
        bytes_to_send[15] = (message.get_origin_id() & 0x000000ff)

        bytes_to_send[16] = (message.get_timestamps() & 0xff00000000000000) >> 56
        bytes_to_send[17] = (message.get_timestamps() & 0x00ff000000000000) >> 48
        bytes_to_send[18] = (message.get_timestamps() & 0x0000ff0000000000) >> 40
        bytes_to_send[19] = (message.get_timestamps() & 0x000000ff00000000) >> 32
        bytes_to_send[20] = (message.get_timestamps() & 0x00000000ff000000) >> 24
        bytes_to_send[21] = (message.get_timestamps() & 0x0000000000ff0000) >> 16
        bytes_to_send[22] = (message.get_timestamps() & 0x000000000000ff00) >> 8
        bytes_to_send[23] = (message.get_timestamps() & 0x00000000000000ff)

        bytes_to_send[24] = (message.get_dlc() & 0xff000000) >> 24
        bytes_to_send[25] = (message.get_dlc() & 0x00ff0000) >> 16
        bytes_to_send[26] = (message.get_dlc() & 0x0000ff00) >> 8
        bytes_to_send[27] = (message.get_dlc() & 0x000000ff)

        for x in range(28, self.global_parameter.get_global_datagram_size()):
            bytes_to_send[x] = 0

        return bytes_to_send

    def send_heartbeat_message(self):
        if DEBUG:
            print("send_heartbeat_message")
        heartbeat_message = LienaHeartbeatMessage(LIENA_SESSION_MANAGEMENT_HEARTBEAT_MESSAGE,
                                                  0,
                                                  0,
                                                  round((time.time() % 86400) * 1000000),
                                                  0)
        datagram = LienaDatagram(self.global_parameter.get_global_datagram_size(), self.encode_heartbeat_message(heartbeat_message))
        self.output_queue.append(datagram)

    def launch(self):
        self.flag = True
        self.heartbeat_thread.start()

    def set_output_cache(self, output_message_cache):
        self.outputMessageCache = output_message_cache

    def terminate(self):
        self.output_queue.clear()
        self.flag = False

    def freeze(self):
        self.stand_by = True

    def enable(self):
        self.stand_by = False

File: src/cxx-dev/LiENa/lienaRealTime/lienaHeartBeatTask.cpp
#include "lienaHeartBeatTask.h"


/**
 * @brief lienaHeartBeatTask::lienaHeartBeatTask
 * @param outputQueue
 * @param globalParameter
 */
lienaHeartBeatTask::lienaHeartBeatTask(lienaOutputQueue* outputQueue, lienaGlobal *globalParameter)
{
    this->outputQueue = outputQueue;
    this->flag = true;
    this->standby = true;
    this->globalParameter = globalParameter;
}

//! ------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaHeartBeatTask::stop
//!
void lienaHeartBeatTask::stop(){
    this->flag = false;
}

//! ------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaHeartBeatTask::launch
//!
void lienaHeartBeatTask::launch(){
    this->standby = false;
    this->start();
}

//! ------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaHeartBeatTask::enable
//!
void lienaHeartBeatTask::enable(){
    this->standby = false;
}

//! ------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaHeartBeatTask::freeze
//!
void lienaHeartBeatTask::freeze(){
    this->standby = true;
}

//! ------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaHeartBeatTask::sendHeartBeatMessage
//!
void lienaHeartBeatTask::sendHeartBeatMessage(){
    uint64_t message_id = LIENA_SESSION_MANAGEMENT_HEARTBEAT_MESSAGE;

    if(DEBUG){
        qDebug()<<"heartbeatMessage | message_id: "<<message_id;
    }
    lienaHeartBeatMessage* heartBeatMessage = new lienaHeartBeatMessage(message_id, 0, 0, 123456, 10);
    lienaDatagram* datagram =this->transformHeartbeatMessageToIgtDatagram(heartBeatMessage);

    this->outputQueue->append(datagram);
}

//! ------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaHeartBeatTask::run
//!
void lienaHeartBeatTask::run(){

    while (this->flag) {
        if(this->standby){
            sleep(1);
            continue;
        }
        this->sendHeartBeatMessage();
        sleep(1);
    }
    return;
}

//! ------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaHeartBeatTask::transformHandShakeCommitMessageToIgtDatagram
//! \param heartBeatMessage
//! \return
//!
lienaDatagram* lienaHeartBeatTask::transformHeartbeatMessageToIgtDatagram(lienaHeartBeatMessage* heartBeatMessage){
    lienaDatagram *datagram = nullptr;

    if(heartBeatMessage != nullptr){

        uint32_t globalDatagramSize = globalParameter->getGlobalDatagramSize();
        char* byteArray = new char[globalDatagramSize];

        uint64_t messageId   = heartBeatMessage->getMessageId();
        uint32_t targetId    = heartBeatMessage->getTargetId();
        uint32_t originId    = heartBeatMessage->getOriginId();
        uint64_t timestampes = heartBeatMessage->getTimestampes();
        uint32_t dlc         = heartBeatMessage->getDLC();

        //-------------------------------message id-----------------------
        byteArray[0] = (uchar)((0xff00000000000000&messageId)>>56);
        byteArray[1] = (uchar)((0x00ff000000000000&messageId)>>48);
        byteArray[2] = (uchar)((0x0000ff0000000000&messageId)>>40);
        byteArray[3] = (uchar)((0x000000ff00000000&messageId)>>32);
        byteArray[4] = (uchar)((0x00000000ff000000&messageId)>>24);
        byteArray[5] = (uchar)((0x0000000000ff0000&messageId)>>16);
        byteArray[6] = (uchar)((0x000000000000ff00&messageId)>>8);
        byteArray[7] = (uchar)((0x00000000000000ff&messageId));

        //-------------------------------target id-----------------------
        byteArray[8] =  (uchar)((0xff000000&targetId)>>24);
        byteArray[9] =  (uchar)((0x00ff0000&targetId)>>16);
        byteArray[10] = (uchar)((0x0000ff00&targetId)>>8);
        byteArray[11] = (uchar)((0x000000ff&targetId));

        //-------------------------------origin id-----------------------
        byteArray[12] =  (uchar)((0xff000000&originId)>>24);
        byteArray[13] =  (uchar)((0x00ff0000&originId)>>16);
        byteArray[14] = (uchar)((0x0000ff00&originId)>>8);
        byteArray[15] = (uchar)((0x000000ff&originId));

        //-------------------------------time stamps -----------------------
        byteArray[16] = (uchar)((0xff00000000000000&timestampes)>>56);
        byteArray[17] = (uchar)((0x00ff000000000000&timestampes)>>48);
        byteArray[18] = (uchar)((0x0000ff0000000000&timestampes)>>40);
        byteArray[19] = (uchar)((0x000000ff00000000&timestampes)>>32);
        byteArray[20] = (uchar)((0x00000000ff000000&timestampes)>>24);
        byteArray[21] = (uchar)((0x0000000000ff0000&timestampes)>>16);
        byteArray[22] = (uchar)((0x000000000000ff00&timestampes)>>8);
        byteArray[23] = (uchar)((0x00000000000000ff&timestampes));

        //-------------------------------dlc -----------------------------

        byteArray[24] =  (uchar)((0xff000000&dlc)>>24);
        byteArray[25] =  (uchar)((0x00ff0000&dlc)>>16);
        byteArray[26] = (uchar)((0x0000ff00&dlc)>>8);
        byteArray[27] = (uchar)((0x000000ff&dlc));

        //-------------------------------body -----------------------

        for(uint32_t i =28; i< globalDatagramSize; i++){
            byteArray[i] = '\0';
        }

        datagram = new lienaDatagram(globalDatagramSize, &byteArray);
    }

    return datagram;
}

